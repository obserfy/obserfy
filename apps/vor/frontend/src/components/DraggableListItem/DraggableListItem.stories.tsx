import { FC, Fragment } from "react"
import { borderBottom, borderRight } from "../../border"
import { useMoveDraggableItem } from "../../hooks/useMoveDraggableItem"
import DraggableListItem from "./DraggableListItem"

export default {
  title: "Core/DraggableListItem",
  component: DraggableListItem,
  parameters: {
    componentSubtitle: "Just a simple DraggableListItem",
  },
}

export const Basic: FC = () => {
  const [items, moveItem] = useMoveDraggableItem([
    { order: 0, id: "asdf" },
    { order: 1, id: "asjfda" },
    { order: 2, id: "ajrhgflkjenbrgk" },
  ])

  return (
    <Fragment>
      {items.map((item) => (
        <DraggableListItem
          key={item.id}
          item={item}
          moveItem={moveItem}
          height={54}
          containerSx={{
            ...borderBottom,
            ...borderRight,
            borderRightColor: "textPrimary",
            borderRightWidth: 2,
            alignItems: "center",
            "&:hover": {
              backgroundColor: "primaryLightest",
            },
          }}
        >
          {item.id}
        </DraggableListItem>
      ))}
    </Fragment>
  )
}
