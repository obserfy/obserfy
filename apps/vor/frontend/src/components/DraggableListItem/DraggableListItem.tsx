import { disableBodyScroll, enableBodyScroll } from "body-scroll-lock"
import { FC, MouseEvent, TouchEvent, useRef, useState } from "react"
import { Box, Flex, ThemeUIStyleObject } from "theme-ui"
import { OrderedItem } from "../../hooks/useMoveDraggableItem"
import { ReactComponent as GridIcon } from "../../icons/grid_round.svg"

function useTransientState<T>(defaultState: T, subscribe: (state: T) => void) {
  const currentState = useRef<T>(defaultState)

  return (state: T) => {
    currentState.current = state
    subscribe(state)
  }
}

interface Props {
  item: OrderedItem
  height: number
  containerSx?: ThemeUIStyleObject
  moveItem: (item: OrderedItem, newOrder: number) => void
  onDrop?: () => void
  disableDrag?: boolean
}

const DraggableListItem: FC<Props> = ({
  children,
  moveItem,
  height,
  containerSx,
  item,
  onDrop,
  disableDrag,
}) => {
  const container = useRef<HTMLDivElement>(null)
  const setTranslateY = useTransientState(0, (state) => {
    if (container.current) {
      container.current.style.transform = `translateY(${state}px)`
    }
  })

  const [isGrabbed, setIsGrabbed] = useState(false)
  const mouseStartingYPosition = useRef(0)
  const originalOrder = useRef(0)
  const dragHandle = useRef<HTMLDivElement>(null)

  // used to make drag handle on the middle of the item
  const itemMidPoint = height / 2

  const recalculateYTranslate = (mouseYPosition: number) => {
    setTranslateY(mouseYPosition - itemMidPoint)
  }

  const handleDrag = (e: MouseEvent | TouchEvent, mouseYPosition: number) => {
    if (!isGrabbed) return
    e.preventDefault()
    e.stopPropagation()

    recalculateYTranslate(mouseYPosition)
    const dragYOffset = mouseYPosition - mouseStartingYPosition.current
    const orderOffset =
      dragYOffset < 0
        ? Math.ceil((dragYOffset - itemMidPoint) / height)
        : Math.floor((dragYOffset + itemMidPoint) / height)

    moveItem(item, originalOrder.current + orderOffset)
  }

  const handleGrabbed = (
    e: MouseEvent | TouchEvent,
    mouseYPosition: number
  ) => {
    if (!dragHandle.current || disableDrag) return
    e.preventDefault()
    e.stopPropagation()
    disableBodyScroll(dragHandle.current, {
      reserveScrollBarGap: true,
    })

    recalculateYTranslate(mouseYPosition)
    mouseStartingYPosition.current = mouseYPosition
    originalOrder.current = item.order
    setIsGrabbed(true)
  }

  const handleReleased = (e: MouseEvent | TouchEvent) => {
    e.preventDefault()
    e.stopPropagation()
    if (!dragHandle.current) return
    enableBodyScroll(dragHandle.current)
    if (onDrop) {
      onDrop()
    }
    setTranslateY(0)
    setIsGrabbed(false)
  }

  return (
    <Box
      sx={{ width: "100%", maxWidth: "inherit", height, ...containerSx }}
      onMouseMove={(e) => handleDrag(e, e.clientY)}
      onTouchMove={(e) => handleDrag(e, e.targetTouches[0].clientY)}
    >
      <Flex
        ref={container}
        backgroundColor={isGrabbed ? "surface" : "transparent"}
        sx={{
          height,
          alignItems: "center",
          userSelect: "none",
          width: "inherit",
          maxWidth: "inherit",
          boxShadow: isGrabbed ? "low" : undefined,
          transition: "background-color .1s ease-in, box-shadow .1s ease-in",
          position: isGrabbed ? "fixed" : "relative",
          zIndex: isGrabbed ? 10 : 1,
          borderRadius: isGrabbed ? "default" : 0,
          top: 0,
        }}
      >
        <Box
          px={3}
          py={2}
          sx={{ cursor: "move" }}
          onClick={(e) => e.preventDefault()}
          onMouseDown={(e) => handleGrabbed(e, e.clientY)}
          onTouchStart={(e) => handleGrabbed(e, e.targetTouches[0].clientY)}
          onMouseUp={handleReleased}
          onTouchEnd={handleReleased}
          onContextMenu={(e) => {
            e.preventDefault()
            e.stopPropagation()
          }}
          ref={dragHandle}
        >
          <GridIcon width={20} />
        </Box>
        {children}
      </Flex>
    </Box>
  )
}

export default DraggableListItem
