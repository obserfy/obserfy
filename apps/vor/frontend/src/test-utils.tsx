/* eslint-disable import/no-extraneous-dependencies */
import { FC, ReactElement } from "react"
import { QueryClient, QueryClientProvider } from "react-query"
import { ThemeProvider, Box, Theme } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import { I18nProvider } from "@lingui/react"
import { i18n } from "@lingui/core"
import { en } from "make-plural/plurals"
import theme from "./gatsby-plugin-theme-ui"
// @ts-ignore
import enCatalog from "../i18n/lingui/en/messages.js"

// any used here because the catalogs are generated b lingui
// not our responsibility
i18n.loadLocaleData("en", { plurals: en })
i18n.load("en", enCatalog.messages as any)
i18n.activate("en")

const queryClient = new QueryClient()
const history = createHistory(createMemorySource("/"))

const AllProviders: FC = ({ children }) => {
  return (
    <I18nProvider i18n={i18n}>
      <QueryClientProvider client={queryClient}>
        <LocationProvider history={history}>
          <ThemeProvider theme={theme as Theme}>
            <Box sx={{ fontSize: [16, 20] }}>{children}</Box>
          </ThemeProvider>
        </LocationProvider>
      </QueryClientProvider>
    </I18nProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
