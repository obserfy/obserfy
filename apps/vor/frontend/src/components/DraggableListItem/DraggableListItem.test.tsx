import { render } from "../../test-utils"
import { Basic } from "./DraggableListItem.stories"

describe("DraggableListItem", () => {
  it("should render correctly", () => {
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
