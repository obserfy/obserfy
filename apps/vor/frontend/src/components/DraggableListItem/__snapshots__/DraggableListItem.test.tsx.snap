// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggableListItem should render correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width: 40em) {
  .emotion-0 {
    font-size: 20px;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  max-width: inherit;
  height: 54px;
  border-bottom-width: 1px;
  border-bottom-style: solid;
  border-bottom-color: var(--theme-ui-colors-border);
  border-right-width: 2px;
  border-right-style: solid;
  border-right-color: var(--theme-ui-colors-textPrimary);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1:hover {
  background-color: var(--theme-ui-colors-primaryLightest);
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 54px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: inherit;
  max-width: inherit;
  -webkit-transition: background-color .1s ease-in,box-shadow .1s ease-in;
  transition: background-color .1s ease-in,box-shadow .1s ease-in;
  position: relative;
  z-index: 1;
  border-radius: 0;
  top: 0;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  cursor: move;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          <div
            width="20"
          />
        </div>
        asdf
      </div>
    </div>
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          <div
            width="20"
          />
        </div>
        asjfda
      </div>
    </div>
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          <div
            width="20"
          />
        </div>
        ajrhgflkjenbrgk
      </div>
    </div>
  </div>
</div>
`;
